### PR TITLE
Density Matrix Construction with GPU MAGMA

### PR DESCRIPTION
**Pull request by robagl9695**
Added the construction of the Density Matrix on the GPU employing MAGMA BLAS subroutines for complex and real cases.

### CODE REVIEW

**Review by bhourahine — COMMENTED** [review]
Could you add an entry to the CHANGELOG file, plus any appropriate edit to the AUTHORS.RST file.
Let me know if the kind(Int64) suggestions from @aradi  works, as this would make the changes a little simpler (and should work for any standard compliant compiler, instead of needing customisation for each one separately).

File: src/dftbp/dftb/densitymatrix.F90 [hunk 1]
```diff
@@ -446,4 +455,277 @@ end subroutine makeDensityMtxCplxBlacs
...
+    endif
+
+    call magmaf_queue_destroy( queue );
+
+  end subroutine makeDensityMtxRealGPU
+
+  !> Make a regular density matrix for the complex wave-function case in the GPU
+  !> Note: In order to save memory, the eigenvectors (which should be intent(in) parameters) are
+  !> overwritten and then restored again
+
```
Comment (bhourahine) [inline src/dftbp/dftb/densitymatrix.F90:599]:
```suggestion
```

File: src/dftbp/dftb/densitymatrix.F90 [hunk 2]
```diff
@@ -446,4 +455,277 @@ end subroutine makeDensityMtxCplxBlacs
...
+#:if WITH_MAGMA
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+!!!!!  MAGMA GPU routines    !!!!!                                                   
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+  !> Make a regular density matrix for the real wave-function case in the GPU
+  !> Note: In order to save memory, the eigenvectors (which should be intent(in) parameters) are
+  !> overwritten and then restored again
+
```
Comment (bhourahine) [inline src/dftbp/dftb/densitymatrix.F90:467]: Doxygen might have problems with the blank line between the comment and the associated routine
```suggestion
```

**Review by bhourahine — COMMENTED** [review]
Just some trivial source formatting stuff

**Comment by bhourahine** [comment]
I've just tested the regressions with intel 2021, and ctest -R GPU fails for the SiC_512_GPU case (but not the non-self-consistent one). I think this is an issue with the complex density matrix routine (as running it with a real matrix works, but making the problem slightly complex fails).

**Comment by robagl9695** [comment]
Resubmitting pull request to include only real valued DM.

**Comment by bhourahine** [comment]
Is the plan to open a new pull request? If so, could you squash the changes into one commit and also re-base on the main repository (git rebase in both cases).

**Comment by robagl9695** [comment]
Hi Ben,

Yes, that's the idea. We have been trying to fix the convergence issues for
complex-valued DM.

However, real-valued DM cases work correctly. We plan to open a new pull
request for just the real-valued DM modification and then another pull
request with the complex-valued, once we fix it.

I will follow your instructions to do so.

Kind regards,

Roberto G.

**Pull request closed without merging**